File: nanopdf-js/src/easy.ts
/**
 * Easy API - Simplified, ergonomic API for common PDF tasks
 *
 * This module provides a high-level, user-friendly interface for the most common
 * PDF operations, with automatic resource management and intuitive method chaining.
 *
 * @example
 * ```typescript
 * import { EasyPDF } from 'nanopdf/easy';
 *
 * // Simple text extraction
 * const text = await EasyPDF.extractText('document.pdf');
 *
 * // Render to PNG
 * const buffer = await EasyPDF.renderPage('document.pdf', 0, { dpi: 300 });
 *
 * // Chain operations
 * await EasyPDF.open('input.pdf')
 *   .getMetadata()
 *   .extractText()
 *   .renderPages({ dpi: 150 })
 *   .save('output.pdf');
 * ```
 */

import { Document } from './document.js';
import { Matrix } from './geometry.js';
import { Colorspace } from './colorspace.js';
import { Pixmap } from './pixmap.js';
import { Context } from './context.js';
import type { RenderOptions } from './render-options.js';
import { dpiToScale } from './render-options.js';
import * as fs from 'fs/promises';

/**
 * Options for rendering pages
 */
export interface EasyRenderOptions {
  /** DPI for rendering (default: 72) */
  dpi?: number;
  /** Width in pixels (alternative to DPI) */
  width?: number;
  /** Height in pixels (alternative to DPI) */
  height?: number;
  /** Colorspace (default: RGB) */
  colorspace?: 'gray' | 'rgb' | 'cmyk';
  /** Include alpha channel (default: false) */
  alpha?: boolean;
  /** Image format for output (default: 'png') */
  format?: 'png' | 'pnm' | 'pam' | 'pbm';
}

/**
 * PDF metadata
 */
export interface PdfMetadata {
  title?: string;
  author?: string;
  subject?: string;
  keywords?: string;
  creator?: string;
  producer?: string;
  creationDate?: Date;
  modDate?: Date;
}

/**
 * Page information
 */
export interface PageInfo {
  pageNumber: number;
  width: number;
  height: number;
  rotation: number;
}

/**
 * Document information
 */
export interface DocumentInfo {
  pageCount: number;
  metadata: PdfMetadata;
  isEncrypted: boolean;
  hasXfa: boolean;
  pages: PageInfo[];
}

/**
 * Text extraction result
 */
export interface ExtractedText {
  text: string;
  pageNumber: number;
  blocks?: Array<{
    text: string;
    bbox: { x: number; y: number; width: number; height: number };
  }>;
}

/**
 * Search result
 */
export interface SearchResult {
  text: string;
  pageNumber: number;
  bbox: { x: number; y: number; width: number; height: number };
}

/**
 * EasyPDF - Fluent builder for PDF operations
 */
export class EasyPDF {
  private doc: Document | null = null;
  private ctx: Context;
  private autoClose = true;

  private constructor(doc: Document) {
    this.doc = doc;
    this.ctx = Context.getDefault();
  }

  /**
   * Open a PDF document
   *
   * @param path - Path to PDF file
   * @param password - Optional password for encrypted PDFs
   * @returns EasyPDF instance for chaining
   *
   * @example
   * ```typescript
   * const pdf = EasyPDF.open('document.pdf');
   * ```
   */
  static open(path: string, password?: string): EasyPDF {
    const doc = Document.open(path, password);
    return new EasyPDF(doc);
  }

  /**
   * Open a PDF from a buffer
   *
   * @param buffer - Buffer containing PDF data
   * @param password - Optional password for encrypted PDFs
   * @returns EasyPDF instance for chaining
   *
   * @example
   * ```typescript
   * const buffer = await fs.readFile('document.pdf');
   * const pdf = EasyPDF.fromBuffer(buffer);
   * ```
   */
  static fromBuffer(buffer: Buffer, password?: string): EasyPDF {
    const doc = Document.fromBuffer(buffer, password);
    return new EasyPDF(doc);
  }

  /**
   * Extract text from a PDF file (static helper)
   *
   * @param path - Path to PDF file
   * @param pageNumber - Optional page number (0-indexed), or undefined for all pages
   * @returns Extracted text
   *
   * @example
   * ```typescript
   * // Extract all text
   * const allText = await EasyPDF.extractText('document.pdf');
   *
   * // Extract from specific page
   * const pageText = await EasyPDF.extractText('document.pdf', 0);
   * ```
   */
  static async extractText(path: string, pageNumber?: number): Promise<string> {
    const pdf = EasyPDF.open(path);
    try {
      if (pageNumber !== undefined) {
        return pdf.extractPageText(pageNumber);
      }
      return pdf.extractAllText();
    } finally {
      pdf.close();
    }
  }

  /**
   * Render a specific page to an image buffer (static helper)
   *
   * @param path - Path to PDF file
   * @param pageNumber - Page number (0-indexed)
   * @param options - Render options
   * @returns Image buffer
   *
   * @example
   * ```typescript
   * const pngBuffer = await EasyPDF.renderPage('document.pdf', 0, {
   *   dpi: 300,
   *   format: 'png'
   * });
   * await fs.writeFile('page.png', pngBuffer);
   * ```
   */
  static async renderPage(
    path: string,
    pageNumber: number,
    options: EasyRenderOptions = {}
  ): Promise<Buffer> {
    const pdf = EasyPDF.open(path);
    try {
      return pdf.renderToBuffer(pageNumber, options);
    } finally {
      pdf.close();
    }
  }

  /**
   * Get document information (static helper)
   *
   * @param path - Path to PDF file
   * @returns Document information
   *
   * @example
   * ```typescript
   * const info = await EasyPDF.getInfo('document.pdf');
   * console.log(`Pages: ${info.pageCount}`);
   * console.log(`Title: ${info.metadata.title}`);
   * ```
   */
  static async getInfo(path: string): Promise<DocumentInfo> {
    const pdf = EasyPDF.open(path);
    try {
      return pdf.getInfo();
    } finally {
      pdf.close();
    }
  }

  /**
   * Search for text in a PDF (static helper)
   *
   * @param path - Path to PDF file
   * @param query - Text to search for
   * @param pageNumber - Optional page number (0-indexed), or undefined for all pages
   * @returns Array of search results
   *
   * @example
   * ```typescript
   * const results = await EasyPDF.search('document.pdf', 'important');
   * console.log(`Found ${results.length} occurrences`);
   * ```
   */
  static async search(
    path: string,
    query: string,
    pageNumber?: number
  ): Promise<SearchResult[]> {
    const pdf = EasyPDF.open(path);
    try {
      return pdf.search(query, pageNumber);
    } finally {
      pdf.close();
    }
  }

  /**
   * Get the number of pages
   *
   * @returns Page count
   */
  get pageCount(): number {
    this.ensureOpen();
    return this.doc!.pageCount;
  }

  /**
   * Check if document is encrypted
   *
   * @returns True if encrypted
   */
  get isEncrypted(): boolean {
    this.ensureOpen();
    return this.doc!.needsPassword();
  }

  /**
   * Get document metadata
   *
   * @returns Metadata object
   *
   * @example
   * ```typescript
   * const metadata = pdf.getMetadata();
   * console.log(metadata.title);
   * console.log(metadata.author);
   * ```
   */
  getMetadata(): PdfMetadata {
    this.ensureOpen();
    const doc = this.doc!;

    const parseDate = (dateStr: string | null): Date | undefined => {
      if (!dateStr) return undefined;
      try {
        // PDF date format: D:YYYYMMDDHHmmSSOHH'mm
        const match = dateStr.match(/^D:(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})/);
        if (match) {
          const [, year, month, day, hour, min, sec] = match;
          return new Date(
            parseInt(year),
            parseInt(month) - 1,
            parseInt(day),
            parseInt(hour),
            parseInt(min),
            parseInt(sec)
          );
        }
        return new Date(dateStr);
      } catch {
        return undefined;
      }
    };

    return {
      title: doc.getMetadata('Title') || undefined,
      author: doc.getMetadata('Author') || undefined,
      subject: doc.getMetadata('Subject') || undefined,
      keywords: doc.getMetadata('Keywords') || undefined,
      creator: doc.getMetadata('Creator') || undefined,
      producer: doc.getMetadata('Producer') || undefined,
      creationDate: parseDate(doc.getMetadata('CreationDate')),
      modDate: parseDate(doc.getMetadata('ModDate'))
    };
  }

  /**
   * Get document information
   *
   * @returns Complete document information
   *
   * @example
   * ```typescript
   * const info = pdf.getInfo();
   * console.log(`Document has ${info.pageCount} pages`);
   * info.pages.forEach((page, i) => {
   *   console.log(`Page ${i}: ${page.width}x${page.height}`);
   * });
   * ```
   */
  getInfo(): DocumentInfo {
    this.ensureOpen();
    const doc = this.doc!;

    const pages: PageInfo[] = [];
    for (let i = 0; i < doc.pageCount; i++) {
      const page = doc.loadPage(i);
      try {
        const bounds = page.bounds();
        pages.push({
          pageNumber: i,
          width: bounds.width,
          height: bounds.height,
          rotation: 0 // TODO: Get rotation from page
        });
      } finally {
        page.drop();
      }
    }

    return {
      pageCount: doc.pageCount,
      metadata: this.getMetadata(),
      isEncrypted: doc.needsPassword(),
      hasXfa: false, // TODO: Check for XFA forms
      pages
    };
  }

  /**
   * Extract text from a specific page
   *
   * @param pageNumber - Page number (0-indexed)
   * @returns Extracted text
   *
   * @example
   * ```typescript
   * const text = pdf.extractPageText(0);
   * console.log(text);
   * ```
   */
  extractPageText(pageNumber: number): string {
    this.ensureOpen();
    const page = this.doc!.loadPage(pageNumber);
    try {
      return page.extractText();
    } finally {
      page.drop();
    }
  }

  /**
   * Extract text from all pages
   *
   * @param separator - Separator between pages (default: '\n\n---\n\n')
   * @returns Extracted text
   *
   * @example
   * ```typescript
   * const allText = pdf.extractAllText();
   * ```
   */
  extractAllText(separator = '\n\n---\n\n'): string {
    this.ensureOpen();
    const texts: string[] = [];

    for (let i = 0; i < this.doc!.pageCount; i++) {
      texts.push(this.extractPageText(i));
    }

    return texts.join(separator);
  }

  /**
   * Extract structured text with bounding boxes
   *
   * @param pageNumber - Page number (0-indexed), or undefined for all pages
   * @returns Array of extracted text with structure
   *
   * @example
   * ```typescript
   * const structured = pdf.extractStructuredText(0);
   * structured.forEach(block => {
   *   console.log(`Text: ${block.text}`);
   *   console.log(`Position: ${block.bbox.x}, ${block.bbox.y}`);
   * });
   * ```
   */
  extractStructuredText(pageNumber?: number): ExtractedText[] {
    this.ensureOpen();
    const results: ExtractedText[] = [];

    const processPage = (pNum: number) => {
      const page = this.doc!.loadPage(pNum);
      try {
        const stext = page.extractStructuredText();
        const blocks = stext.blocks.map((block) => ({
          text: block.text || '',
          bbox: {
            x: block.bbox.x,
            y: block.bbox.y,
            width: block.bbox.width,
            height: block.bbox.height
          }
        }));

        results.push({
          text: page.extractText(),
          pageNumber: pNum,
          blocks
        });
      } finally {
        page.drop();
      }
    };

    if (pageNumber !== undefined) {
      processPage(pageNumber);
    } else {
      for (let i = 0; i < this.doc!.pageCount; i++) {
        processPage(i);
      }
    }

    return results;
  }

  /**
   * Search for text in the document
   *
   * @param query - Text to search for
   * @param pageNumber - Optional page number (0-indexed), or undefined for all pages
   * @returns Array of search results
   *
   * @example
   * ```typescript
   * const results = pdf.search('important');
   * results.forEach(result => {
   *   console.log(`Found on page ${result.pageNumber}: ${result.text}`);
   * });
   * ```
   */
  search(query: string, pageNumber?: number): SearchResult[] {
    this.ensureOpen();
    const results: SearchResult[] = [];

    const searchPage = (pNum: number) => {
      const page = this.doc!.loadPage(pNum);
      try {
        const hits = page.searchText(query);
        hits.forEach((hit) => {
          results.push({
            text: query,
            pageNumber: pNum,
            bbox: {
              x: hit.x,
              y: hit.y,
              width: hit.width,
              height: hit.height
            }
          });
        });
      } finally {
        page.drop();
      }
    };

    if (pageNumber !== undefined) {
      searchPage(pageNumber);
    } else {
      for (let i = 0; i < this.doc!.pageCount; i++) {
        searchPage(i);
      }
    }

    return results;
  }

  /**
   * Render a page to a buffer
   *
   * @param pageNumber - Page number (0-indexed)
   * @param options - Render options
   * @returns Image buffer
   *
   * @example
   * ```typescript
   * const pngBuffer = pdf.renderToBuffer(0, { dpi: 300 });
   * await fs.writeFile('page.png', pngBuffer);
   * ```
   */
  renderToBuffer(pageNumber: number, options: EasyRenderOptions = {}): Buffer {
    this.ensureOpen();
    const page = this.doc!.loadPage(pageNumber);

    try {
      // Calculate transform matrix
      let matrix = Matrix.identity();

      if (options.dpi) {
        const scale = dpiToScale(options.dpi);
        matrix = Matrix.scale(scale, scale);
      } else if (options.width || options.height) {
        const bounds = page.bounds();
        const scaleX = options.width ? options.width / bounds.width : 1;
        const scaleY = options.height ? options.height / bounds.height : 1;
        const scale = Math.min(scaleX, scaleY);
        matrix = Matrix.scale(scale, scale);
      }

      // Determine colorspace
      let colorspace: Colorspace;
      switch (options.colorspace) {
        case 'gray':
          colorspace = Colorspace.deviceGray();
          break;
        case 'cmyk':
          colorspace = Colorspace.deviceCMYK();
          break;
        case 'rgb':
        default:
          colorspace = Colorspace.deviceRGB();
          break;
      }

      // Render to pixmap
      const pixmap = page.toPixmap(matrix, colorspace, options.alpha ?? false);

      try {
        // Convert to buffer
        const format = options.format || 'png';
        return pixmap.toBuffer(format);
      } finally {
        pixmap.drop();
      }
    } finally {
      page.drop();
    }
  }

  /**
   * Render a page to a file
   *
   * @param pageNumber - Page number (0-indexed)
   * @param outputPath - Output file path
   * @param options - Render options
   *
   * @example
   * ```typescript
   * await pdf.renderToFile(0, 'page.png', { dpi: 300 });
   * ```
   */
  async renderToFile(
    pageNumber: number,
    outputPath: string,
    options: EasyRenderOptions = {}
  ): Promise<void> {
    const buffer = this.renderToBuffer(pageNumber, options);
    await fs.writeFile(outputPath, buffer);
  }

  /**
   * Render all pages
   *
   * @param options - Render options
   * @returns Array of image buffers
   *
   * @example
   * ```typescript
   * const pages = pdf.renderAll({ dpi: 150 });
   * pages.forEach((buffer, i) => {
   *   fs.writeFileSync(`page-${i}.png`, buffer);
   * });
   * ```
   */
  renderAll(options: EasyRenderOptions = {}): Buffer[] {
    this.ensureOpen();
    const buffers: Buffer[] = [];

    for (let i = 0; i < this.doc!.pageCount; i++) {
      buffers.push(this.renderToBuffer(i, options));
    }

    return buffers;
  }

  /**
   * Render all pages to files
   *
   * @param outputPattern - Output file pattern (use {page} for page number)
   * @param options - Render options
   *
   * @example
   * ```typescript
   * await pdf.renderAllToFiles('output/page-{page}.png', { dpi: 150 });
   * ```
   */
  async renderAllToFiles(
    outputPattern: string,
    options: EasyRenderOptions = {}
  ): Promise<void> {
    this.ensureOpen();

    for (let i = 0; i < this.doc!.pageCount; i++) {
      const outputPath = outputPattern.replace('{page}', i.toString());
      await this.renderToFile(i, outputPath, options);
    }
  }

  /**
   * Disable automatic closing (for manual resource management)
   *
   * @returns this for chaining
   */
  keepOpen(): this {
    this.autoClose = false;
    return this;
  }

  /**
   * Close the document and free resources
   *
   * @example
   * ```typescript
   * const pdf = EasyPDF.open('document.pdf');
   * try {
   *   // Use pdf...
   * } finally {
   *   pdf.close();
   * }
   * ```
   */
  close(): void {
    if (this.doc) {
      this.doc.close();
      this.doc = null;
    }
  }

  /**
   * Use the PDF document within a callback, automatically closing it
   *
   * @param callback - Callback function
   * @returns Result of callback
   *
   * @example
   * ```typescript
   * const text = await EasyPDF.open('document.pdf').use(pdf => {
   *   return pdf.extractAllText();
   * });
   * ```
   */
  use<T>(callback: (pdf: this) => T): T {
    try {
      return callback(this);
    } finally {
      if (this.autoClose) {
        this.close();
      }
    }
  }

  /**
   * Use the PDF document within an async callback, automatically closing it
   *
   * @param callback - Async callback function
   * @returns Promise with result of callback
   *
   * @example
   * ```typescript
   * const info = await EasyPDF.open('document.pdf').useAsync(async pdf => {
   *   return pdf.getInfo();
   * });
   * ```
   */
  async useAsync<T>(callback: (pdf: this) => Promise<T>): Promise<T> {
    try {
      return await callback(this);
    } finally {
      if (this.autoClose) {
        this.close();
      }
    }
  }

  private ensureOpen(): void {
    if (!this.doc) {
      throw new Error('PDF document is closed');
    }
  }
}

/**
 * Namespace for quick utility functions
 */
export namespace PDFUtils {
  /**
   * Extract text from a PDF file
   */
  export const extractText = EasyPDF.extractText;

  /**
   * Render a page to buffer
   */
  export const renderPage = EasyPDF.renderPage;

  /**
   * Get document information
   */
  export const getInfo = EasyPDF.getInfo;

  /**
   * Search for text
   */
  export const search = EasyPDF.search;
}

File: go-nanopdf/easy/easy.go
// Package easy provides a simplified, ergonomic API for common PDF tasks.
//
// This package offers a high-level interface with automatic resource management
// and intuitive method chaining, making PDF operations simple and safe.
//
// Example usage:
//
//	import "github.com/lexmata/nanopdf/go-nanopdf/easy"
//
//	// Extract text
//	text, err := easy.ExtractText("document.pdf")
//
//	// Render to PNG
//	err := easy.RenderToPNG("document.pdf", "output.png", 0, 300)
//
//	// Fluent API
//	pdf, _ := easy.Open("document.pdf")
//	defer pdf.Close()
//	info := pdf.GetInfo()
//	text := pdf.ExtractAllText()
//
package easy

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/lexmata/nanopdf/go-nanopdf"
)

// RenderOptions specifies options for rendering pages
type RenderOptions struct {
	// DPI for rendering (default: 72)
	DPI float32
	// Width in pixels (alternative to DPI)
	Width int
	// Height in pixels (alternative to DPI)
	Height int
	// Colorspace (default: "rgb")
	Colorspace string // "gray", "rgb", "cmyk"
	// Include alpha channel (default: false)
	Alpha bool
	// Image format for output (default: "png")
	Format string // "png", "pnm", "pam", "pbm"
}

// Metadata represents PDF metadata
type Metadata struct {
	Title        string
	Author       string
	Subject      string
	Keywords     string
	Creator      string
	Producer     string
	CreationDate *time.Time
	ModDate      *time.Time
}

// PageInfo represents information about a page
type PageInfo struct {
	PageNumber int
	Width      float32
	Height     float32
	Rotation   int
}

// DocumentInfo represents complete document information
type DocumentInfo struct {
	PageCount   int
	Metadata    Metadata
	IsEncrypted bool
	Pages       []PageInfo
}

// SearchResult represents a text search result
type SearchResult struct {
	Text       string
	PageNumber int
	BBox       nanopdf.Rect
}

// PDF is a fluent builder for PDF operations
type PDF struct {
	doc       *nanopdf.Document
	autoClose bool
}

// Open opens a PDF document
//
//	pdf, err := easy.Open("document.pdf")
//	if err != nil {
//	    return err
//	}
//	defer pdf.Close()
func Open(path string) (*PDF, error) {
	doc, err := nanopdf.OpenDocument(path, "")
	if err != nil {
		return nil, err
	}

	return &PDF{
		doc:       doc,
		autoClose: true,
	}, nil
}

// OpenWithPassword opens a password-protected PDF document
func OpenWithPassword(path, password string) (*PDF, error) {
	doc, err := nanopdf.OpenDocument(path, password)
	if err != nil {
		return nil, err
	}

	return &PDF{
		doc:       doc,
		autoClose: true,
	}, nil
}

// FromBytes opens a PDF from byte data
func FromBytes(data []byte) (*PDF, error) {
	doc, err := nanopdf.OpenDocumentFromMemory(data, "")
	if err != nil {
		return nil, err
	}

	return &PDF{
		doc:       doc,
		autoClose: true,
	}, nil
}

// ============================================================================
// Static Helper Functions (convenience methods)
// ============================================================================

// ExtractText extracts all text from a PDF file
//
//	text, err := easy.ExtractText("document.pdf")
//	if err != nil {
//	    return err
//	}
//	fmt.Println(text)
func ExtractText(path string) (string, error) {
	pdf, err := Open(path)
	if err != nil {
		return "", err
	}
	defer pdf.Close()

	return pdf.ExtractAllText()
}

// ExtractPageText extracts text from a specific page
//
//	text, err := easy.ExtractPageText("document.pdf", 0)
func ExtractPageText(path string, pageNumber int) (string, error) {
	pdf, err := Open(path)
	if err != nil {
		return "", err
	}
	defer pdf.Close()

	return pdf.ExtractPageText(pageNumber)
}

// GetPageCount returns the number of pages in a PDF
//
//	count, err := easy.GetPageCount("document.pdf")
func GetPageCount(path string) (int, error) {
	pdf, err := Open(path)
	if err != nil {
		return 0, err
	}
	defer pdf.Close()

	return pdf.PageCount(), nil
}

// GetMetadata returns PDF metadata
//
//	meta, err := easy.GetMetadata("document.pdf")
//	fmt.Println(meta.Title)
func GetMetadata(path string) (*Metadata, error) {
	pdf, err := Open(path)
	if err != nil {
		return nil, err
	}
	defer pdf.Close()

	return pdf.GetMetadata(), nil
}

// GetInfo returns complete document information
//
//	info, err := easy.GetInfo("document.pdf")
//	fmt.Printf("Document has %d pages\n", info.PageCount)
func GetInfo(path string) (*DocumentInfo, error) {
	pdf, err := Open(path)
	if err != nil {
		return nil, err
	}
	defer pdf.Close()

	return pdf.GetInfo(), nil
}

// RenderToPNG renders a page to PNG
//
//	err := easy.RenderToPNG("document.pdf", "output.png", 0, 300)
func RenderToPNG(pdfPath, outputPath string, pageNumber int, dpi float32) error {
	pdf, err := Open(pdfPath)
	if err != nil {
		return err
	}
	defer pdf.Close()

	return pdf.RenderToFile(pageNumber, outputPath, RenderOptions{DPI: dpi, Format: "png"})
}

// RenderAllToPNG renders all pages to PNG files
//
//	err := easy.RenderAllToPNG("document.pdf", "output/page-{page}.png", 150)
func RenderAllToPNG(pdfPath, outputPattern string, dpi float32) error {
	pdf, err := Open(pdfPath)
	if err != nil {
		return err
	}
	defer pdf.Close()

	return pdf.RenderAllToFiles(outputPattern, RenderOptions{DPI: dpi, Format: "png"})
}

// Search searches for text in a PDF
//
//	results, err := easy.Search("document.pdf", "important")
//	fmt.Printf("Found %d occurrences\n", len(results))
func Search(pdfPath, query string) ([]int, error) {
	pdf, err := Open(pdfPath)
	if err != nil {
		return nil, err
	}
	defer pdf.Close()

	results, err := pdf.Search(query, -1)
	if err != nil {
		return nil, err
	}

	// Extract unique page numbers
	pageSet := make(map[int]bool)
	for _, result := range results {
		pageSet[result.PageNumber] = true
	}

	pages := make([]int, 0, len(pageSet))
	for page := range pageSet {
		pages = append(pages, page)
	}

	return pages, nil
}

// IsEncrypted checks if a PDF is encrypted
//
//	if encrypted, _ := easy.IsEncrypted("document.pdf"); encrypted {
//	    fmt.Println("PDF is password protected")
//	}
func IsEncrypted(path string) (bool, error) {
	pdf, err := Open(path)
	if err != nil {
		// If error mentions password, it's encrypted
		if strings.Contains(err.Error(), "password") {
			return true, nil
		}
		return false, err
	}
	defer pdf.Close()

	return pdf.IsEncrypted(), nil
}

// SaveTextToFile extracts text and saves to a file
//
//	err := easy.SaveTextToFile("document.pdf", "output.txt")
func SaveTextToFile(pdfPath, outputPath string) error {
	text, err := ExtractText(pdfPath)
	if err != nil {
		return err
	}

	return os.WriteFile(outputPath, []byte(text), 0644)
}

// QuickSummary returns a human-readable summary of the PDF
//
//	summary, err := easy.QuickSummary("document.pdf")
//	fmt.Println(summary)
func QuickSummary(path string) (string, error) {
	pdf, err := Open(path)
	if err != nil {
		return "", err
	}
	defer pdf.Close()

	info := pdf.GetInfo()

	var builder strings.Builder
	builder.WriteString(fmt.Sprintf("Document: %s\n", filepath.Base(path)))
	builder.WriteString(fmt.Sprintf("Pages: %d\n", info.PageCount))

	if info.Metadata.Title != "" {
		builder.WriteString(fmt.Sprintf("Title: %s\n", info.Metadata.Title))
	}
	if info.Metadata.Author != "" {
		builder.WriteString(fmt.Sprintf("Author: %s\n", info.Metadata.Author))
	}
	if info.Metadata.Subject != "" {
		builder.WriteString(fmt.Sprintf("Subject: %s\n", info.Metadata.Subject))
	}

	builder.WriteString(fmt.Sprintf("Encrypted: %v\n", info.IsEncrypted))

	return builder.String(), nil
}

// ============================================================================
// PDF Methods
// ============================================================================

// PageCount returns the number of pages
func (p *PDF) PageCount() int {
	return p.doc.PageCount()
}

// IsEncrypted returns whether the document is encrypted
func (p *PDF) IsEncrypted() bool {
	return p.doc.NeedsPassword()
}

// GetMetadata returns PDF metadata
func (p *PDF) GetMetadata() *Metadata {
	meta := &Metadata{
		Title:    p.doc.GetMetadata("Title"),
		Author:   p.doc.GetMetadata("Author"),
		Subject:  p.doc.GetMetadata("Subject"),
		Keywords: p.doc.GetMetadata("Keywords"),
		Creator:  p.doc.GetMetadata("Creator"),
		Producer: p.doc.GetMetadata("Producer"),
	}

	// Parse dates if present
	if creationDate := p.doc.GetMetadata("CreationDate"); creationDate != "" {
		if t, err := parsePDFDate(creationDate); err == nil {
			meta.CreationDate = &t
		}
	}
	if modDate := p.doc.GetMetadata("ModDate"); modDate != ""{
		if t, err := parsePDFDate(modDate); err == nil {
			meta.ModDate = &t
		}
	}

	return meta
}

// GetInfo returns complete document information
func (p *PDF) GetInfo() *DocumentInfo {
	pageCount := p.PageCount()
	pages := make([]PageInfo, pageCount)

	for i := 0; i < pageCount; i++ {
		page, err := p.doc.LoadPage(i)
		if err != nil {
			continue
		}

		bounds := page.Bounds()
		pages[i] = PageInfo{
			PageNumber: i,
			Width:      bounds.Width(),
			Height:     bounds.Height(),
			Rotation:   0, // TODO: Get rotation from page
		}

		page.Drop()
	}

	return &DocumentInfo{
		PageCount:   pageCount,
		Metadata:    *p.GetMetadata(),
		IsEncrypted: p.IsEncrypted(),
		Pages:       pages,
	}
}

// ExtractPageText extracts text from a specific page
func (p *PDF) ExtractPageText(pageNumber int) (string, error) {
	page, err := p.doc.LoadPage(pageNumber)
	if err != nil {
		return "", err
	}
	defer page.Drop()

	return page.ExtractText(), nil
}

// ExtractAllText extracts text from all pages
func (p *PDF) ExtractAllText() (string, error) {
	return p.ExtractAllTextWithSeparator("\n\n---\n\n")
}

// ExtractAllTextWithSeparator extracts text from all pages with a custom separator
func (p *PDF) ExtractAllTextWithSeparator(separator string) (string, error) {
	var builder strings.Builder
	pageCount := p.PageCount()

	for i := 0; i < pageCount; i++ {
		text, err := p.ExtractPageText(i)
		if err != nil {
			return "", err
		}

		builder.WriteString(text)
		if i < pageCount-1 {
			builder.WriteString(separator)
		}
	}

	return builder.String(), nil
}

// Search searches for text in the document
func (p *PDF) Search(query string, pageNumber int) ([]SearchResult, error) {
	var results []SearchResult

	searchPage := func(pNum int) error {
		page, err := p.doc.LoadPage(pNum)
		if err != nil {
			return err
		}
		defer page.Drop()

		hits := page.SearchText(query)
		for _, hit := range hits {
			results = append(results, SearchResult{
				Text:       query,
				PageNumber: pNum,
				BBox:       hit,
			})
		}

		return nil
	}

	if pageNumber >= 0 {
		if err := searchPage(pageNumber); err != nil {
			return nil, err
		}
	} else {
		for i := 0; i < p.PageCount(); i++ {
			if err := searchPage(i); err != nil {
				return nil, err
			}
		}
	}

	return results, nil
}

// RenderToBytes renders a page to an image buffer
func (p *PDF) RenderToBytes(pageNumber int, opts RenderOptions) ([]byte, error) {
	page, err := p.doc.LoadPage(pageNumber)
	if err != nil {
		return nil, err
	}
	defer page.Drop()

	// Calculate transform matrix
	matrix := nanopdf.MatrixIdentity()

	if opts.DPI > 0 {
		scale := opts.DPI / 72.0
		matrix = nanopdf.ScaleMatrix(scale, scale)
	} else if opts.Width > 0 || opts.Height > 0 {
		bounds := page.Bounds()
		scaleX := float32(1.0)
		scaleY := float32(1.0)

		if opts.Width > 0 {
			scaleX = float32(opts.Width) / bounds.Width()
		}
		if opts.Height > 0 {
			scaleY = float32(opts.Height) / bounds.Height()
		}

		scale := scaleX
		if opts.Height > 0 && scaleY < scaleX {
			scale = scaleY
		}

		matrix = nanopdf.ScaleMatrix(scale, scale)
	}

	// Render to pixmap
	pixmap, err := page.ToPixmap(matrix, nil, opts.Alpha)
	if err != nil {
		return nil, err
	}
	defer pixmap.Drop()

	// Convert to format
	format := opts.Format
	if format == "" {
		format = "png"
	}

	return pixmap.ToBytes(format)
}

// RenderToFile renders a page to a file
func (p *PDF) RenderToFile(pageNumber int, outputPath string, opts RenderOptions) error {
	data, err := p.RenderToBytes(pageNumber, opts)
	if err != nil {
		return err
	}

	return os.WriteFile(outputPath, data, 0644)
}

// RenderAll renders all pages
func (p *PDF) RenderAll(opts RenderOptions) ([][]byte, error) {
	results := make([][]byte, p.PageCount())

	for i := 0; i < p.PageCount(); i++ {
		data, err := p.RenderToBytes(i, opts)
		if err != nil {
			return nil, err
		}
		results[i] = data
	}

	return results, nil
}

// RenderAllToFiles renders all pages to files
func (p *PDF) RenderAllToFiles(outputPattern string, opts RenderOptions) error {
	for i := 0; i < p.PageCount(); i++ {
		outputPath := strings.Replace(outputPattern, "{page}", fmt.Sprintf("%d", i), -1)

		// Create directory if needed
		dir := filepath.Dir(outputPath)
		if err := os.MkdirAll(dir, 0755); err != nil {
			return err
		}

		if err := p.RenderToFile(i, outputPath, opts); err != nil {
			return err
		}
	}

	return nil
}

// KeepOpen disables automatic closing
func (p *PDF) KeepOpen() *PDF {
	p.autoClose = false
	return p
}

// Close closes the document and frees resources
func (p *PDF) Close() error {
	if p.doc != nil {
		p.doc.Close()
		p.doc = nil
	}
	return nil
}

// Use executes a callback with the PDF, automatically closing it
func (p *PDF) Use(callback func(*PDF) error) error {
	defer func() {
		if p.autoClose {
			p.Close()
		}
	}()

	return callback(p)
}

// ============================================================================
// Helper Functions
// ============================================================================

// parsePDFDate parses a PDF date string
func parsePDFDate(dateStr string) (time.Time, error) {
	// PDF date format: D:YYYYMMDDHHmmSSOHH'mm
	if !strings.HasPrefix(dateStr, "D:") {
		return time.Time{}, errors.New("invalid PDF date format")
	}

	dateStr = strings.TrimPrefix(dateStr, "D:")

	// Parse at least YYYYMMDD
	if len(dateStr) < 8 {
		return time.Time{}, errors.New("date string too short")
	}

	// Extract components
	year := dateStr[0:4]
	month := dateStr[4:6]
	day := dateStr[6:8]

	hour := "00"
	minute := "00"
	second := "00"

	if len(dateStr) >= 10 {
		hour = dateStr[8:10]
	}
	if len(dateStr) >= 12 {
		minute = dateStr[10:12]
	}
	if len(dateStr) >= 14 {
		second = dateStr[12:14]
	}

	dateString := fmt.Sprintf("%s-%s-%sT%s:%s:%sZ", year, month, day, hour, minute, second)
	return time.Parse(time.RFC3339, dateString)
}

